### PR TITLE
Don't notify cache about nonces; pass the account nonce provider, instead

### PR DESCRIPTION
**Pull request by andreibancioiu**
## Reasoning behind the pull request
- This is an improvement upon https://github.com/multiversx/mx-chain-go/pull/6581.
- In certain exotic cases, concurrent notifications about account nonces might have lead the mempool to hold stale account nonces (very minor issue, but still an issue).
- The flow of _forgetting_ all account nonces after reverting a block is suboptimal (and somehow over-engineered).
- Transaction de-duplication (based on nonces) wasn't entirely correct.
- Accounts with transactions recently proposed in scheduled blocks had some issues with following transactions (selection was delayed due to missing nonce notifications).
  
## Proposed changes
- https://github.com/multiversx/mx-chain-storage-go/pull/57
- Also merged this one (PR was previously lost): https://github.com/multiversx/mx-chain-go/pull/6338
- Also merged this one: https://github.com/multiversx/mx-chain-go/pull/6604
- Drop `txs.accountTxsShards`
- Fix old concurrency issue in `requestHandler` (found by @sstanculeanu and @miiu96): https://github.com/multiversx/mx-chain-go/pull/6603/commits/5bb1b40eae211c4da0131c1931c1edc2cd3d993f.

## Testing procedure
- Standard testing
- Heavy load, to trigger evictions from the mempool

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by sasurobert — DISMISSED** [review]

**Review by raduchis — APPROVED** [review]

File: state/accountsDB.go [hunk 1]
```diff
@@ -653,9 +649,7 @@ func (adb *AccountsDB) GetExistingAccount(address []byte) (vmcommon.AccountHandl
 		return nil, fmt.Errorf("%w in GetExistingAccount", ErrNilAddress)
 	}
 
-	log.Trace("accountsDB.GetExistingAccount",
-		"address", hex.EncodeToString(address),
-	)
+	log.Trace("accountsDB.GetExistingAccount", "address", address)
```
Comment (andreibancioiu) [inline state/accountsDB.go:652]: Optimizations.

File: node/external/transactionAPI/apiTransactionProcessor.go [hunk 2]
```diff
@@ -401,8 +401,11 @@ func (atp *apiTransactionProcessor) getFieldGettersForTx(wrappedTx *txcache.Wrap
 	guardedTx, isGuardedTx := wrappedTx.Tx.(data.GuardedTransactionHandler)
 	if isGuardedTx {
 		fieldGetters[signatureField] = hex.EncodeToString(guardedTx.GetSignature())
-		fieldGetters[guardianField] = atp.addressPubKeyConverter.SilentEncode(guardedTx.GetGuardianAddr(), log)
-		fieldGetters[guardianSignatureField] = hex.EncodeToString(guardedTx.GetGuardianSignature())
+
+		if len(guardedTx.GetGuardianAddr()) > 0 {
```
Comment (andreibancioiu) [inline node/external/transactionAPI/apiTransactionProcessor.go:405]: This is from https://github.com/multiversx/mx-chain-go/pull/6604.

File: integrationTests/multiShard/relayedTx/relayedTx_test.go [hunk 3]
```diff
@@ -188,7 +188,7 @@ func testRelayedTransactionInMultiShardEnvironmentWithSmartContractTX(
 		}
 		time.Sleep(time.Second)
 
-		roundToPropagateMultiShard := int64(25)
+		roundToPropagateMultiShard := int64(40)
```
Comment (andreibancioiu) [inline integrationTests/multiShard/relayedTx/relayedTx_test.go:191]: Test was extremely hard to debug.

It seems that it was affected by the new behavior of the cache, that does not return transactions with lower nonces during selection.

File: dataRetriever/requestHandlers/requestHandler.go [hunk 4]
```diff
@@ -571,10 +571,12 @@ func (rrh *resolverRequestHandler) RequestValidatorInfo(hash []byte) {
 		return
 	}
 
+	epoch := rrh.getEpoch()
```
Comment (andreibancioiu) [inline dataRetriever/requestHandlers/requestHandler.go:574]: CC: @sstanculeanu

File: genesis/process/shardGenesisBlockCreator.go [hunk 5]
```diff
@@ -600,6 +600,7 @@ func createProcessorsForShardGenesisBlock(arg ArgsGenesisBlockCreator, enableEpo
 		disabledScheduledTxsExecutionHandler,
 		disabledProcessedMiniBlocksTracker,
 		arg.TxExecutionOrderHandler,
+		disabledGuardian.NewDisabledGuardedAccountHandler(),
```
Comment (andreibancioiu) [inline genesis/process/shardGenesisBlockCreator.go:603]: Genesis components, we can pass this object.

File: process/block/preprocess/miniBlockBuilder.go [hunk 6]
```diff
@@ -136,15 +130,6 @@ func checkMiniBlocksBuilderArgs(args miniBlocksBuilderArgs) error {
 	return nil
 }
 
-func (mbb *miniBlocksBuilder) updateAccountShardsInfo(tx *transaction.Transaction, wtx *txcache.WrappedTransaction) {
```
Comment (andreibancioiu) [inline process/block/preprocess/miniBlockBuilder.go:139]: Was used for mempool notifications.

File: process/constants.go [hunk 7]
```diff
@@ -143,3 +144,6 @@ const TxCacheSelectionGasRequested = 10_000_000_000
 
 // TxCacheSelectionMaxNumTxs defines the maximum number of transactions that should be selected from the cache.
 const TxCacheSelectionMaxNumTxs = 50000
+
+// TxCacheSelectionLoopMaximumDuration defines the maximum duration for the loop that selects transactions from the cache.
+const TxCacheSelectionLoopMaximumDuration = 250 * time.Millisecond
```
Comment (andreibancioiu) [inline process/constants.go:149]: Question for review: all right?

**Review by raduchis — CHANGES_REQUESTED** [review]

File: go.mod [hunk 8]
```diff
@@ -20,7 +20,7 @@ require (
 	github.com/multiversx/mx-chain-es-indexer-go v1.7.10
 	github.com/multiversx/mx-chain-logger-go v1.0.15
 	github.com/multiversx/mx-chain-scenario-go v1.4.4
-	github.com/multiversx/mx-chain-storage-go v1.0.17
+	github.com/multiversx/mx-chain-storage-go v1.0.18-0.20241124203643-93c958ad66cf
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply go.mod:23]: Will be set after all merges into the feat branch, right before preparing the release.

File: process/block/preprocess/basePreProcess.go [hunk 9]
```diff
@@ -120,6 +120,7 @@ type basePreProcess struct {
 	blockSizeComputation       BlockSizeComputationHandler
 	balanceComputation         BalanceComputationHandler
 	accounts                   state.AccountsAdapter
+	accountStateProvider       *accountStateProvider
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply process/block/preprocess/basePreProcess.go:123]: It's already handled a bit differently in the next PR:

https://github.com/multiversx/mx-chain-go/pull/6628/files#diff-acfb4fa3ec74315de16957c120913b7150aaeb7ccd1374dc1b12ce0c4a4d8048R1414

File: process/block/preprocess/transactions.go [hunk 10]
```diff
@@ -39,11 +38,6 @@ const selectionGasBandwidthIncreasePercent = 200
 // 130% to allow 30% overshooting estimations for scheduled SC calls
 const selectionGasBandwidthIncreaseScheduledPercent = 130
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply process/block/preprocess/transactions.go:39]: Changed :+1:

File: process/block/preprocess/transactions.go [hunk 11]
```diff
@@ -34,15 +33,10 @@ var _ process.PreProcessor = (*transactions)(nil)
 var log = logger.GetOrCreate("process/block/preprocess")
 
 // 200% bandwidth to allow 100% overshooting estimations
-const selectionGasBandwidthIncreasePercent = 200
+const selectionGasBandwidthIncreasePercent = 400
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply process/block/preprocess/transactions.go:36]: We should, yes. Added MX-16201 (we have some other constants to move to config).

File: go.mod [hunk 12]
```diff
@@ -20,7 +20,7 @@ require (
 	github.com/multiversx/mx-chain-es-indexer-go v1.7.10
 	github.com/multiversx/mx-chain-logger-go v1.0.15
 	github.com/multiversx/mx-chain-scenario-go v1.4.4
-	github.com/multiversx/mx-chain-storage-go v1.0.17
+	github.com/multiversx/mx-chain-storage-go v1.0.18-0.20241126200205-8a0cb502901f
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply go.mod:23]: We reference the latest commit from here: https://github.com/multiversx/mx-chain-storage-go/pull/57/commits.

A proper tag be set after all merges into the feat branch, right before preparing the release.

File: dataRetriever/txpool/memorytests/memory_test.go [hunk 13]
```diff
@@ -36,18 +35,18 @@ func TestShardedTxPool_MemoryFootprint(t *testing.T) {
 
 	journals = append(journals, runScenario(t, newScenario(200, 1, core.MegabyteSize, "0"), memoryAssertion{200, 200}, memoryAssertion{0, 1}))
 	journals = append(journals, runScenario(t, newScenario(10, 1000, 20480, "0"), memoryAssertion{190, 205}, memoryAssertion{1, 4}))
-	journals = append(journals, runScenario(t, newScenario(10000, 1, 1024, "0"), memoryAssertion{10, 16}, memoryAssertion{4, 10}))
-	journals = append(journals, runScenario(t, newScenario(1, 60000, 256, "0"), memoryAssertion{30, 40}, memoryAssertion{10, 16}))
-	journals = append(journals, runScenario(t, newScenario(10, 10000, 100, "0"), memoryAssertion{36, 52}, memoryAssertion{16, 24}))
+	journals = append(journals, runScenario(t, newScenario(10000, 1, 1024, "0"), memoryAssertion{10, 16}, memoryAssertion{0, 10}))
```
[the remark this answers is not in the document]
Comment (raduchis) [reply dataRetriever/txpool/memorytests/memory_test.go:38]: ok!

File: process/block/preprocess/transactions.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply process/block/preprocess/transactions.go]: ok!

File: process/block/preprocess/basePreProcess.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply process/block/preprocess/basePreProcess.go:123]: ok

**Review by raduchis — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**